java/ffi/src/main/java/sleeper/foreign/FFIBridge.java

```diff
@@ -0,0 +1,118 @@
+     * unpack them to a temporary directory.
+     */
+    private static final JniExtractor EXTRACTOR = NativeLoader.getJniExtractor();
+
+    /** Paths in the JAR file where a native library may have been placed. */
+    private static final String[] LIB_PATHS = {
+        "natives/x86_64-unknown-linux-gnu/release", "natives/aarch64-unknown-linux-gnu/release",
+        // Rust debug builds will place libraries in different locations
+        "natives/x86_64-unknown-linux-gnu/debug", "natives/aarch64-unknown-linux-gnu/debug"};
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FFIBridge.class);
+
+    /**
+     * Attempt to load the native foreign function library.
+     *
+     * The native library will be extracted from the classpath and unpacked to a temporary
+     * directory. The library is then loaded and linked. Multiple locations are checked in the
+     * classpath, representing different architectures. Thus, if we attempt to load a library for
+     * the wrong CPU architecture, loading will fail and the next path will be tried. This way, we
+     * maintain a single JAR file that can work across multiple CPU architectures.
+     *
+     * Each call to this library will extract and link a new copy of the foreign library so clients
+     * are encouraged to cache results of successful calls to this function.
+     *
+     * @param  clazz       the interface describing the foreign function calls
+     * @param  <T>         interface type containing Java functions stubs
+     * @return             the native call interface
+     * @throws IOException if an error occurs during loading or linking the native library
+     */
+    public static synchronized <T extends ForeignFunctions> T createForeignInterface(Class<T> clazz) throws IOException {
+        try {
+            return extractAndLink(clazz, "compaction");
+        } catch (UnsatisfiedLinkError err) {
+            throw (IOException) new IOException("Could not initialise foreign library bridge", err);
+        }
+    }
+
+    /**
+     * Loads the named library after extracting it from the classpath.
+     *
+     * This function extracts the named library from a JAR on the classpath and attempts to load it
+     * and bind it to the given interface class. The paths in the array {@link LIB_PATHS} are tried
+     * in order. If a library is found at a path, this method will attempt to load it. If no library
+     * is found on the classpath or it can't be loaded (e.g. wrong binary format), the next path
+     * will be tried.
+     *
+     * The library named should be given without platform prefixes, e.g. "foo" will be expanded into
+     * "libfoo.so" or "foo.dll" as appropriate for this platform.
+     *
+     * @param  <T>                  the type of the interface to the native code
+     * @param  clazz                the class of the interface to the native code
+     * @param  libName              the library name to extract without platform prefixes.
```

Comment:
This can either be capitalised as a sentence or the full stop can be removed.

See the Oracle style guide under `@param`: "The description begins with a lowercase letter if it is a phrase (contains no verb), or an uppercase letter if it is a sentence. End the phrase with a period only if another phrase or sentence follows it."

https://www.oracle.com/uk/technical-resources/articles/java/javadoc-tool.html